rfd/0153-resource-guidelines.md

```diff
@@ -0,0 +1,469 @@
+return foo, nil
+```
+
+#### Get
+
+The `Get` RPC takes the parameters required to match a resource (usually the resource name should suffice), and returns
+the matched resource. If a resource contains secrets (passwords, private keys, etc.) The request should also take
+a `with_secrets` parameter that allows users to request the sensitive information. By default, the secrets should not be
+provided in the response.
+
+The request MUST fail and return a `trace.NotFound` error if there is no matching resource in the backend.
+
+```protobuf
+// Returns a single Foo matching the request
+    rpc GetFoo(GetFooRequest) returns (GetFooResponse);
+
+message GetFooRequest {
+  // A filter to match the Foo by. Some resource may require more parameters to match and
+  // may not use the name at all.
+  string name = 1;
+  // Specifies whether to load associated secrets with the resource.
+  // NOTE: This is only required for resources which have secrets(passwords, 
+  // private keys, mfa devices, etc). If a resource does not have any 
+  // secrets this can be omitted.
+  bool with_secrets = 2;
+}
+
+message GetFooResponse {
+  // The matched Foo resource.
+  Foo foo = 1;
+}
+
+```
+
+#### List
+
+The `List` RPC takes the requested page size and starting point and returns a list of resources that match. If there are
+additional resources, the response MUST also include a token that indicates where the next page of results begins. If a
+resource contains secrets (passwords, private keys, etc.) The request should also take a `with_secrets` parameter that
+allows users to request the sensitive information. By default, the secrets should not be provided in the response.
+
+```protobuf
+// Returns a page of Foo and the token to find the next page of items.
+    rpc ListFoo(ListFooRequest) returns (ListFooResponse);
+
+message ListFoosRequest {
+  // The maximum number of items to return.
+  // The server may impose a different page size at its discretion.
+  int32 page_size = 1;
+  // The next_page_token value returned from a previous List request, if any.
+  string page_token = 2;
+  // Specifies whether to load associated secrets with the resource.
+  // NOTE: This is only required for resources which have secrets(passwords, 
+  // private keys, mfa devices, etc). If a resource does not have any 
+  // secrets this can be omitted.
+  bool with_secrets = 3;
+}
+
+message ListFoosResponse {
+  // The page of Foo that matched the request.
+  repeated Foo foos = 1;
+  // Token to retrieve the next page of results, or empty if there are no
+  // more results in the list.
+  string next_page_token = 2;
+}
+```
+
+A listing operation should not abort entirely if a single item cannot be (un)marshalled, it should instead be logged,
+and the rest of the page should be processed. Aborting an entire page when a single entry is invalid causes the cache
+to be permanently unhealthy since it is never able to initialize loading the affected resource.
```

Comment:
```suggestion
and the rest of the page should be processed. Aborting an entire page when a single entry is invalid causes teleport
to be permanently unhealthy since it is never able to load or cache the affected resource.
```

Cache init is probably the first operation that will show up as broken in logs, but all features that might directly or indirectly rely on consuming the resource set will be permanently borked.